### PR TITLE
Add new `rustc` target for Arm64 machines that can target the iphonesimulator

### LINKED ISSUE #81632: Create a `aarch64-apple-ios-sim` target for `rustc`
Currently in our application we cannot run the iOS simulator on M1 machines because it is expecting `x86_64` architecture. Forcing `arm64` using the `rustc` target `aarch64-apple-ios` can build the project and ship the application but this targets the iOS Device SDK, and will not build in the simulator for testing with this error: 

```
ld: <path/to/lib> building for iOS Simulator, but linking in object file built for iOS, for architecture arm64
```

`rustc` should provide an `aarch-apple-ios-sim` target that uses the iOS Simulator SDK.

### PR DESCRIPTION
**Pull request by deg4uss3r**
This PR lands a new target (`aarch64-apple-ios-sim`) that targets arm64 iphone simulator, previously unreachable from Apple Silicon machines. 

resolves #81632

r? @shepmaster

### CODE REVIEW

File: src/doc/rustc/src/platform-support.md [hunk 1]
```diff
@@ -153,6 +153,7 @@ not available.
 target | std | host | notes
 -------|-----|------|-------
 `aarch64-apple-ios-macabi` | ? |  | Apple Catalyst on ARM64
+`aarch64-apple-ios-sim` | ? |  | Apple iOS Simulator on  ARM64
```
Comment (LingMan) [inline src/doc/rustc/src/platform-support.md:156]:
```suggestion
`aarch64-apple-ios-sim` | ? |  | Apple iOS Simulator on ARM64
```

**Comment by badboy** [comment]
tl;dr: I finally made this work with my static libraries to work inside the iOS simulator on an M1 macbook pro. I had to use the triple `arm64-apple-ios7.0.0-simulator` to make it behave.

heads up: This is based on #81451 to get LLVM 12. I have not tested yet if that's required.

Details on how I got there:

First I looked at what the difference is between C code compiled for arm64 iOS and for arm64 iOS Simulator:

```
❯ cat t.c
#include <stdio.h>

int foo()
{
  printf("hello world\n");
  return 0;
}

❯ clang -target arm64-apple-ios-simulator t.c -c -o t-sim.o -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator.sdk
❯ clang -target arm64-apple-ios t.c -c -o t-ios.o -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS.sdk
❯ otool -l t-sim.o | rg 'VERSION|platform|version'
      cmd LC_BUILD_VERSION
 platform 7
❯ otool -l t-ios.o | rg 'VERSION|platform|version'
      cmd LC_VERSION_MIN_IPHONEOS
  version 7.0
```

As you can see the simulator one uses `LC_BUILD_VERSION`, whereas the iOS one uses `LC_VERSION_MIN_IPHONEOS`.

After a bunch of tries, trying to trace the rustc and LLVM code I was unable to fully determine how LLVM decides which way to use.
No matter what I did the objects falling out of rustc used `LC_VERSION_MIN_IPHONEOS`.

So I went back to look at the C code again.
This time comparing the LLVM IR.

Here's the relevant difference between `t.c` compiled with `clang -emit-llvm` for iOS or iOS-simulator:

```
❯ diff -u t-ios.ll t-sim.ll  | head -8
--- t-ios.ll	2021-02-12 11:10:21.000000000 +0100
+++ t-sim.ll	2021-02-12 11:10:03.000000000 +0100
@@ -1,7 +1,7 @@
 ; ModuleID = 't.c'
 source_filename = "t.c"
 target datalayout = "e-m:o-i64:64-i128:128-n32:64-S128"
-target triple = "arm64-apple-ios7.0.0"
+target triple = "arm64-apple-ios7.0.0-simulator"
```

Low and behold: the triple has the version in it.

So I swapped out the triple in `aarch64_apple_ios_sim.rs`, recompiled my compiler, then used that to build my test library:

```
❯ cat src/lib.rs
#[no_mangle]
pub extern "C" fn hello_world() -> u32 {
    1+1
}
❯ rustc +stage1 --target aarch64-apple-ios-sim src/lib.rs --crate-type staticlib
❯ otool -l liblib.a | rg 'VERSION|platform|version' | head -2
      cmd LC_BUILD_VERSION
 platform 7
```

There we go!

Now it turns out: for macOS we do the same: we put the deployment target version into the triple: https://github.com/rust-lang/rust/blob/a118ee2c13cc96ceb27bd5030c1cca1052377604/compiler/rustc_target/src/spec/apple_base.rs#L57-L65

I guess we need to start doing that for iOS too? I can work on integrating that into the code of this PR.
(Side note: I took my rustc build to my iOS project and I can now run tests using the simulator, there's more work ahead on our side because of course we can't build one fat library anymore for hardware and simulator targets ...)

**Comment by deg4uss3r** [comment]
Awesome detective work @badboy !!!! 

Looks like we'll need LLVM12 still I tried building against the latest `master` with the triple changed to `arm64-apple-ios7.0.0-simulator`

and I'm still getting: 

```
Showing All Messages
In /Users/rth/work/core/target/aarch64-apple-ios-sim/debug/libcore.a(sqlite3.o), building for iOS Simulator, but linking in object file built for iOS, for architecture arm64
```

**Comment by badboy** [comment]
> Awesome detective work @badboy !!!!
> 
> Looks like we'll need LLVM12 still I tried building against the latest `master` with the triple changed to `arm64-apple-ios7.0.0-simulator`
> 
> and I'm still getting:
> 
> ```
> Showing All Messages
> In /Users/rth/work/core/target/aarch64-apple-ios-sim/debug/libcore.a(sqlite3.o), building for iOS Simulator, but linking in object file built for iOS, for architecture arm64
> ```

Well, that one might be because sqlite3 is built with the wrong target. If sqlite3 is built using the `cc` crate you might need a `export CFLAGS_aarch64_apple_ios_sim="-target arm64-apple-ios7.0.0-simulator"` (or `TARGET_CFLAGS="..."`) to make it work.

I should fine some time tomorrow to extend this PR with the necessary steps and also look into `cc` and whether it needs changes to make it work out of the box.

**Comment by deg4uss3r** [comment]
💯 adding that flag to my build script works I can run my app, this is awesome @badboy 😃 

I'll let you add your commits as I am sure it's much cleaner than my branch is now lol

**Comment by badboy** [comment]
not sure if you noticed, but I sent a PR to your branch here: https://github.com/deg4uss3r/rust/pull/1
If you merge that it should show up here and from there it can get proper review.

**Comment by shepmaster** [comment]
Seems believable, and it sounds like two separate people got it to work, so that's about as good of testing as I'd expect for this level.

Would you mind squashing the commits (don't really need the intermediate "fix formatting") and then we can merge!

**Comment by deg4uss3r** [comment]
@shepmaster ungracefully done!

**Comment by LingMan** [comment]
@deg4uss3r You can squash the typo fix, too. I really don't care about credit for something that minor.

**Comment by shepmaster** [comment]
@bors r+ rollup

Thanks!
